File: fb_decoder.py
import sys
import struct
import flatbuffers
import Header, Person, Group, Gender, Type

# event handler interface
class EventHandler:
    def handle_event(self, data, size):
        pass

# Event handler for Header
class HeaderHandler(EventHandler):
    def handle_event(self, data, size):
        header = Header.Header.GetRootAsHeader(data, 0)

        print("--Header--")
        print("Message Length:", header.MsgLen())
        print("Message Type:", header.MsgType())

# Event handler for Person
class PersonHandler(EventHandler):
    def handle_event(self, data, size):
        person = Person.Person.GetRootAsPerson(data, 0)

        print("--Person--")
        print("Person Name:", person.Name().decode('utf-8'))
        print("Person Age:", person.Age())
        print("Person Weight:", person.Weight())
        print("Person Gender:", person.Gender())

# Event handler for Group
class GroupHandler(EventHandler):
    def handle_event(self, data, size):
        group = Group.Group.GetRootAsGroup(data, 0)

        print("--Group--")
        print("Group:", group.GroupName().decode('utf-8'), "Average Age:", group.AverageAge(), "Average Weight:", group.AverageWeight())
        print("Names:", end=' ')
        for i in range(group.NamesLength()):
            print(group.Names(i).decode('utf-8'), ", " , end=' ')
        print()

def main():
    if len(sys.argv) != 2:
        print("Usage:", sys.argv[0], "<binary_file>")
        return 1

    # Read the binary data from the file
    filename = sys.argv[1]
    try:
        with open(filename, 'rb') as infile:
            header_size = 24
            buffer = infile.read(header_size)
            headerHandler = HeaderHandler()
            personHandler = PersonHandler()
            groupHandler = GroupHandler()

            while buffer:

                header = Header.Header.GetRootAsHeader(buffer, 0)
                headerHandler.handle_event(buffer, header_size)

                message_size = header.MsgLen()
                message_buffer = infile.read(message_size)

                if header.MsgType() == Type.Type.Person:
                    personHandler.handle_event(message_buffer, message_size)
                elif header.MsgType() == Type.Type.Group:
                    groupHandler.handle_event(message_buffer, message_size)

                buffer = infile.read(header_size)

    except FileNotFoundError:
        print("Error: Cannot open file for reading:", filename)
        return 1

if __name__ == "__main__":
    sys.exit(main())

File: fb_decoder.cpp
#include "flatbuffers/flatbuffers.h"
#include "client_generated.h"
#include <iostream>
#include <fstream>
#include <vector>

// Define event handler interface
class EventHandler {
public:
    virtual void HandleEvent(const uint8_t* data, size_t size) = 0;
};

// Event handler for Header
class HeaderHandler : public EventHandler {
public:
    void HandleEvent(const uint8_t* data, size_t size) override {
        flatbuffers::Verifier verifier(data, size);
        if (!VerifyHeaderBuffer(verifier)) {
            std::cerr << "Error: Invalid header in the binary data." << std::endl;
            return;
        }

        auto header = GetHeader(data);

        // Print Header values
        std::cout << "--Header--" << std::endl;
        std::cout << "Message Length: " << header->msg_len() << std::endl;
        std::cout << "Message Type: " << EnumNameType(header->msg_type()) << std::endl;
    }
};

// Event handler for Person
class PersonHandler : public EventHandler {
public:
    void HandleEvent(const uint8_t* data, size_t size) override {
        const Person* person = flatbuffers::GetRoot<Person>(data);

        std::cout << "--Person--" << std::endl;
        std::cout << "Person Name: " << person->name()->c_str() << std::endl;
        std::cout << "Person Age: " << person->age() << std::endl;
        std::cout << "Person Weight: " << person->weight() << std::endl;
        std::cout << "Person Gender: " << EnumNameGender(person->gender()) << std::endl;
    }
};

// Event handler for Group
class GroupHandler : public EventHandler {
public:
    void HandleEvent(const uint8_t* data, size_t size) override {
        const Group* group = flatbuffers::GetRoot<Group>(data);

        std::cout << "--Group--" << std::endl;
        std::cout << "Group: " << group->group_name()->c_str() << ", Average Age: " << group->average_age()
            << ", Average Weight: " << group->average_weight() << std::endl;
        std::cout << "Names: ";
        for (size_t j = 0; j < group->names()->size(); ++j) {
            std::cout << group->names()->Get(j)->c_str() << ", ";
        }
        std::cout << std::endl;
    }
};

int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <binary_file>" << std::endl;
        return 1;
    }

    // Read the binary data from the file
    const char* filename = argv[1];
    std::ifstream infile(filename, std::ios::binary);
    if (!infile) {
        std::cerr << "Error: Cannot open file for reading: " << filename << std::endl;
        return 1;
    }

    int32_t header_size = 24;
    std::vector<uint8_t> buffer(header_size);

    // Create event handlers
    HeaderHandler headerHandler;
    PersonHandler personHandler;
    GroupHandler groupHandler;

    while (true) {
        // Read the entire binary data into a buffer
        infile.read(reinterpret_cast<char*>(buffer.data()), header_size);

        if (infile.eof()) {
            std::cerr << "Reached the end of the file." << std::endl;
            break;
        }

        // Dispatch events based on message type
        auto header = GetHeader(buffer.data());
        headerHandler.HandleEvent(buffer.data(), header_size);

        // Read the actual message data
        size_t message_size = header->msg_len();
        std::vector<uint8_t> message_buffer(message_size);
        infile.read(reinterpret_cast<char*>(message_buffer.data()), message_size);

        if (header->msg_type() == Type_Person) {
            personHandler.HandleEvent(message_buffer.data(), message_size);
        } else if (header->msg_type() == Type_Group) {
            groupHandler.HandleEvent(message_buffer.data(), message_size);
        }
    }

    infile.close();

    return 0;
}

